webdriver-spec.html

```diff
@@ -5132,44 +5137,58 @@
       <li><p>Jump to the last step in this overall set of steps.
    </ol>
 
-   <li><p>Let <var>undo actions</var> be equal to the <a>current
-    session</a>'s <a>input cancel list</a> where the <a>action
-    type</a> is <code>"key"</code> in reverse order.
-
-   <li><p><a>Dispatch tick actions</a> with arguments <var>undo
-    actions</var> and <var>duration</var> <code>0</code>.
-
-   <li><p>Remove all <var>undo actions</var> from the <a>current
-    session</a>'s <a>input cancel list</a>.
+   <li><p>Let <var>keyboard</var> be a new <a>key input source</a>.
 
-  <li><p>Let <var>i</var> be 0.
+   <li><p>Let <var>undo actions</var> be an empty list.
 
-  <li><p>Let <var>length</var> be the length of <var>text</var>.
-
-  <li><p>While <var>i</var> &lt; <var>length</var>:
+   <li><p>For each <var>grapheme cluster</var> in <var>text</var>:
 
     <ol>
-      <li><p>Let <var>char</var> be the entry in <var>text</var> at
-        index <var>i</var>.
-
-      <li><p>Let <var>key to type</var> be the result of <a>processing keys</a>
-        with argument <var>key</var>.
-
-      <li><p>If the <var>key to type</var> is a <a>modifier key</a> then set the relevant
-        attribute on <a>Keyboard Event</a> to true for the rest of this loop.
-
-      <li><p>Generate <a data-lt=keyDown-Event>keyDown</a> <a>Keyboard Event</a>
-        with the relevant <a>keyboard event code</a> for <var>key to type</var>
-        against <var>active element</var>.
+     <li><p>Let <var>key to type</var> be the result of <a>processing
+      keys</a> with argument <var>grapheme cluster</var>.
```

Comment:
I think maybe this needs to change later. If the intention is that Shift + 2 should produce @, we need to establish some internal state here _across_ grapheme clusters.

(This isn’t an issue.)